demos/signature/EVP_DSA_Signature_demo.c

```diff
@@ -0,0 +1,236 @@
+    "And by opposing, end them, to die to sleep;\n"
+    "No more, and by a sleep, to say we end\n"
+    "The heart-ache, and the thousand natural shocks\n"
+    "That flesh is heir to? tis a consumation\n"
+;
+
+static const char ALG[] = "DSA";
+static const char DIGEST[] = "SHA256";
+static const int NUMBITS = 2048;
+static const char * const PROPQUERY = NULL;
+
+static int generate_dsa_params(OSSL_LIB_CTX *libctx,
+                               EVP_PKEY **p_params)
+{
+    int result = 0;
+
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *params = NULL;
+
+    ctx = EVP_PKEY_CTX_new_from_name(libctx, ALG, PROPQUERY);
+    if (ctx == NULL)
+        goto end;
+
+    if (EVP_PKEY_paramgen_init(ctx) <= 0)
+        goto end;
+
+    if (EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx, NUMBITS) <= 0)
+        goto end;
+    if (EVP_PKEY_paramgen(ctx, &params) <= 0)
+        goto end;
+    if (params == NULL)
+        goto end;
+
+    result = 1;
+end:
+    if(!result) {
+        EVP_PKEY_free(params);
+        params = NULL;
+    }
+    EVP_PKEY_CTX_free(ctx);
+    *p_params = params;
+    fprintf(stdout, "Params:\n");
+    EVP_PKEY_print_params_fp(stdout, params, 4, NULL);
+    fprintf(stdout, "\n");
+
+    return result;
+}
+
+static int generate_dsa_key(OSSL_LIB_CTX *libctx,
+                            EVP_PKEY *params,
+                            EVP_PKEY **p_pkey)
+{
+    int result = 0;
+
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, params,
+                                     NULL);
+    if (ctx == NULL)
+        goto end;
+    if (EVP_PKEY_keygen_init(ctx) <= 0)
+        goto end;
+
+    if (EVP_PKEY_keygen(ctx, &pkey) <= 0)
+        goto end;
+    if (pkey == NULL)
+        goto end;
+
+    result = 1;
+end:
+    if(!result) {
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+    }
+    EVP_PKEY_CTX_free(ctx);
+    *p_pkey = pkey;
+    fprintf(stdout, "Generating public/private key pair:\n");
+    EVP_PKEY_print_public_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+    EVP_PKEY_print_private_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+    EVP_PKEY_print_params_fp(stdout, pkey, 4, NULL);
+    fprintf(stdout, "\n");
+
+    return result;
+}
+
+static int demo_sign(OSSL_LIB_CTX *libctx,
+                     size_t *p_sig_len, unsigned char **p_sig_value,
+                     EVP_PKEY *pkey)
+{
+    int result = 0;
+    size_t sig_len = 0;
+    unsigned char *sig_value = NULL;
+    EVP_MD_CTX *ctx = NULL;
+
+    ctx = EVP_MD_CTX_create();
+    if (ctx == NULL)
+        goto end;
+
+    if (EVP_DigestSignInit_ex(ctx, NULL, DIGEST, libctx, NULL, pkey, NULL) != 1)
+        goto end;
+
+   if (EVP_DigestSignUpdate(ctx, hamlet_1, sizeof(hamlet_1)) != 1)
+        goto end;
+
+   if (EVP_DigestSignUpdate(ctx, hamlet_2, sizeof(hamlet_2)) != 1)
+        goto end;
+
+    if (EVP_DigestSignFinal(ctx, NULL, &sig_len) != 1)
+        goto end;
+    if (sig_len <= 0)
```

Comment:
This cant be negative..

Reply:
I'd have preferred the original check.  Who knows what error conditions might be introduced in the future.